tests/client_test.go

```diff
@@ -11,56 +12,89 @@ import (
 )
 
 var tmspType = "socket"
+var tmspAddr = "tcp://127.0.0.1:46659"
 
-func TestClient(t *testing.T) {
+func TestNonPersistent(t *testing.T) {
+	testProcedure(t, tmspAddr, "", 0, false, true)
+}
+
+func TestPersistent(t *testing.T) {
+	dbName := "testDb"
+	os.RemoveAll(dbName) //remove the database if exists for any reason
+	testProcedure(t, tmspAddr, dbName, 0, false, false)
+	testProcedure(t, tmspAddr, dbName, 0, true, true)
+	os.RemoveAll(dbName) //cleanup, remove database that was created by testProcedure
+}
 
-	addr := "tcp://127.0.0.1:46659"
+func testProcedure(t *testing.T, addr, dbName string, cache int, testPersistence, clearRecords bool) {
+
+	checkErr := func(err error) {
+		if err != nil {
+			t.Fatal(err.Error())
+			return
+		}
+	}
 
 	// Start the listener
-	mApp := app.NewMerkleEyesApp()
+	mApp := app.NewMerkleEyesApp(dbName, cache)
 	s, err := server.NewServer(addr, tmspType, mApp)
-	if err != nil {
-		t.Fatal(err.Error())
-		return
-	}
-	defer s.Stop()
+
+	defer func() { //Close the database, and server
+		mApp.CloseDb()
+		s.Stop()
+	}()
+	checkErr(err)
 
 	// Create client
 	cli, err := eyes.NewClient(addr, tmspType)
-	if err != nil {
-		t.Fatal(err.Error())
-		return
+
+	//Overwrite previous defer statement
```

Comment:
Hmm.. maybe we are talking about different things: https://play.golang.org/p/Lpq6q1nMNm

```
package main

import (
	"fmt"
)

func main() {
	defer func() { fmt.Println("third") }()
	defer func() { fmt.Println("second") }()
	defer func() { fmt.Println("first") }()
}

==>>>
first
second
third
```